python/pyspark/sql/connect/session.py

```diff
@@ -215,7 +215,38 @@ def createDataFrame(
         _inferred_schema: Optional[StructType] = None
 
         if isinstance(data, pd.DataFrame):
-            _table = pa.Table.from_pandas(data)
+            from pandas.api.types import (  # type: ignore[attr-defined]
+                is_datetime64_dtype,
+                is_datetime64tz_dtype,
+            )
+            from pyspark.sql.pandas.types import (
+                _check_series_convert_timestamps_internal,
+                _get_local_timezone,
+            )
+
+            # Copying the frame to avoid modifying it.
+            data_copy = data.copy()
```

Comment:
Maybe let's copy if it has `is_datetime64tz_dtype` or `is_datetime64_dtype` (since `pa.Schema.from_pandas` itself copies and converts it)

Reply:
Or would be nicer if we can avoid copying here at all if possible

Reply:
I can modify this to only copy when necessary. The biggest issue is that i need to copy it to modify the pandas type before converting it to an arrow table. I think there is likely a better way of doing that that does not require touching the pandas dataframe. 

The general question is, is it ok to modify the input dataframe or should it be immutable. 

However, even in PySpark we modify the input dataframe when we localize the timezones.

Reply:
The input DataFrame should not change (as you are doing it now). Maybe we should just copy when it has timestamps ...

Reply:
I changed this to only copy when the DF contains timestamps.